crypto/fipsmodule/ec/gfp_p384.c

```diff
@@ -63,120 +65,12 @@ static const BN_ULONG ONE[P384_LIMBS] = {
-   * we divide `two_a` by two (mod q), we need to get the value `2**383`, which
-   * we obviously can't get with just a right shift.
-   *
-   * `q` is odd, and `a` is odd, so `a + q` is even. We could calculate
-   * `(a + q) >> 1` and then reduce it mod `q`. However, then we would have to
-   * keep track of an extra most significant bit. We can avoid that by instead
-   * calculating `(a >> 1) + ((q + 1) >> 1)`. The `1` in `q + 1` is the least
-   * significant bit of `a`. `q + 1` is even, which means it can be shifted
-   * without losing any bits. Since `q` is odd, `q - 1` is even, so the largest
-   * odd field element is `q - 2`. Thus we know that `a <= q - 2`. We know
-   * `(q + 1) >> 1` is `(q + 1) / 2` since (`q + 1`) is even. The value of
-   * `a >> 1` is `(a - 1)/2` since the shift will drop the least significant
-   * bit of `a`, which is 1. Thus:
-   *
-   * sum  =  ((q + 1) >> 1) + (a >> 1)
-   * sum  =  (q + 1)/2 + (a >> 1)       (substituting (q + 1)/2)
-   *     <=  (q + 1)/2 + (q - 2 - 1)/2  (substituting a <= q - 2)
-   *     <=  (q + 1)/2 + (q - 3)/2      (simplifying)
-   *     <=  (q + 1 + q - 3)/2          (factoring out the common divisor)
-   *     <=  (2q - 2)/2                 (simplifying)
-   *     <=  q - 1                      (simplifying)
-   *
-   * Thus, no reduction of the sum mod `q` is necessary. */
-
-  Limb is_odd = constant_time_is_nonzero_w(a[0] & 1);
-
-  /* r = a >> 1. */
-  Limb carry = a[P384_LIMBS - 1] & 1;
-  r[P384_LIMBS - 1] = a[P384_LIMBS - 1] >> 1;
-  for (size_t i = 1; i < P384_LIMBS; ++i) {
-    Limb new_carry = a[P384_LIMBS - i - 1];
-    r[P384_LIMBS - i - 1] =
-        (a[P384_LIMBS - i - 1] >> 1) | (carry << (LIMB_BITS - 1));
-    carry = new_carry;
-  }
-
-  static const Elem Q_PLUS_1_SHR_1 = {
-    TOBN(0x00000000, 0x80000000), TOBN(0x7fffffff, 0x80000000),
-    TOBN(0xffffffff, 0xffffffff), TOBN(0xffffffff, 0xffffffff),
-    TOBN(0xffffffff, 0xffffffff), TOBN(0x7fffffff, 0xffffffff),
-  };
-
-  Elem adjusted;
-  BN_ULONG carry2 = limbs_add(adjusted, r, Q_PLUS_1_SHR_1, P384_LIMBS);
-  dev_assert_secret(carry2 == 0);
-  (void)carry2;
-  copy_conditional(r, adjusted, is_odd);
-}
-
-static inline void elem_mul_mont(Elem r, const Elem a, const Elem b) {
-  static const BN_ULONG Q_N0[] = {
-    BN_MONT_CTX_N0(0x1, 0x1)
-  };
-  /* XXX: Not (clearly) constant-time; inefficient.*/
-  bn_mul_mont(r, a, b, Q, Q_N0, P384_LIMBS);
-}
-
-static inline void elem_mul_by_2(Elem r, const Elem a) {
-  LIMBS_shl_mod(r, a, Q, P384_LIMBS);
-}
-
-static INLINE_IF_POSSIBLE void elem_mul_by_3(Elem r, const Elem a) {
-  /* XXX: inefficient. TODO: Replace with an integrated shift + add. */
-  Elem doubled;
-  elem_add(doubled, a, a);
-  elem_add(r, doubled, a);
-}
-
-static inline void elem_sqr_mont(Elem r, const Elem a) {
-  /* XXX: Inefficient. TODO: Add a dedicated squaring routine. */
-  elem_mul_mont(r, a, a);
-}
+#include "ecp_nistz.inl"
 
 void p384_elem_sub(Elem r, const Elem a, const Elem b) {
```

Comment:
It seems like all, or at least most, of these `extern` functions' definitions are boilerplate that can be moved to what would be gfp.h like the point multiplication/addition/doubling functions.